src/basic/string-util.rs

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+use std::ffi::{CStr, CString};
+use std::os::raw::c_char;
+use std::ptr;
+
+/// Extract the i'nth line from the specified string. Returns > 0 if there are more lines after that,
+/// and == 0 if we are looking at the last line or already beyond the last line. As special
+/// optimization, if the first line is requested and the string only consists of one line we return
+/// NULL, indicating the input string should be used as is, and avoid a memory allocation for a very
+/// common case.
+#[no_mangle]
+pub extern "C" fn string_extract_line(s: *const c_char, i: usize, ret: *mut *mut c_char) -> i32 {
+    if ret.is_null() {
+        panic!("string_extract_line(): 'ret' cannot be NULL");
+    }
+
+    let s = unsafe {
+        CStr::from_ptr(s)
+            .to_string_lossy()
+            .into_owned()
+    };
+    let mut found: bool = false;
+
+    for (j, item) in s.lines().enumerate() {
```

Comment:
what's a "line" in rust btw? i.e. what separators are accepted? \n? \r? \0? combinations thereof?

Reply:
`\n` or `\r\n`

https://doc.rust-lang.org/std/primitive.str.html#method.lines

Reply:
that's quite some difference to our current logic... i'd prefer if we didn#t introduce that silently

Reply:
Well, the tests are passing :-) Apart from the `\r\n`, what else is different?

Reply:
FWIW using `\n` unconditionally is simple enough:
```rust
    for (j, item) in s.split('\n').enumerate() {
```